src/phase0/state_transition.rs

```diff
@@ -46,3 +56,134 @@ pub fn apply_block<
+        MAX_ATTESTER_SLASHINGS,
+        MAX_ATTESTATIONS,
+        MAX_DEPOSITS,
+        MAX_VOLUNTARY_EXITS,
+    >,
+) -> bool {
+    let proposer_index = signed_block.message.proposer_index;
+    let proposer = state
+        .validators
+        .get(proposer_index)
+        .expect("failed to get validator");
+    let signing_root = match compute_signing_root(
+        signed_block.message,
+        get_domain(state, DomainType::BeaconProposer, None),
+    ) {
+        Ok(root) => root,
+        Err(_) => return false,
+    };
+
+    proposer
+        .pubkey
+        .verify_signature(&signing_root.as_ref()[..], signed_block.signature)
+}
+
+pub fn get_domain<
+    const SLOTS_PER_HISTORICAL_ROOT: usize,
+    const HISTORICAL_ROOTS_LIMIT: usize,
+    const ETH1_DATA_VOTES_BOUND: usize,
+    const VALIDATOR_REGISTRY_LIMIT: usize,
+    const EPOCHS_PER_HISTORICAL_VECTOR: usize,
+    const EPOCHS_PER_SLASHINGS_VECTOR: usize,
+    const MAX_VALIDATORS_PER_COMMITTEE: usize,
+    const PENDING_ATTESTATIONS_BOUND: usize,
+>(
+    state: BeaconState<
+        SLOTS_PER_HISTORICAL_ROOT,
+        HISTORICAL_ROOTS_LIMIT,
+        ETH1_DATA_VOTES_BOUND,
+        VALIDATOR_REGISTRY_LIMIT,
+        EPOCHS_PER_HISTORICAL_VECTOR,
+        EPOCHS_PER_SLASHINGS_VECTOR,
+        MAX_VALIDATORS_PER_COMMITTEE,
+        PENDING_ATTESTATIONS_BOUND,
+    >,
+    domain_type: DomainType,
+    epoch: Option<Epoch>,
+) -> Domain {
+    let epoch = if epoch.is_none() {
+        get_current_epoch(state)
+    } else {
+        epoch.unwrap()
+    };
+    let fork_version = if epoch < state.fork.epoch {
+        state.fork.previous_version.clone()
+    } else {
+        state.fork.current_version.clone()
+    };
+
+    return compute_domain(
+        domain_type,
+        fork_version,
+        state.genesis_validators_root.clone(),
+    );
+}
+
+pub fn get_current_epoch<
+    const SLOTS_PER_HISTORICAL_ROOT: usize,
+    const HISTORICAL_ROOTS_LIMIT: usize,
+    const ETH1_DATA_VOTES_BOUND: usize,
+    const VALIDATOR_REGISTRY_LIMIT: usize,
+    const EPOCHS_PER_HISTORICAL_VECTOR: usize,
+    const EPOCHS_PER_SLASHINGS_VECTOR: usize,
+    const MAX_VALIDATORS_PER_COMMITTEE: usize,
+    const PENDING_ATTESTATIONS_BOUND: usize,
+>(
+    state: BeaconState<
+        SLOTS_PER_HISTORICAL_ROOT,
+        HISTORICAL_ROOTS_LIMIT,
+        ETH1_DATA_VOTES_BOUND,
+        VALIDATOR_REGISTRY_LIMIT,
+        EPOCHS_PER_HISTORICAL_VECTOR,
+        EPOCHS_PER_SLASHINGS_VECTOR,
+        MAX_VALIDATORS_PER_COMMITTEE,
+        PENDING_ATTESTATIONS_BOUND,
+    >,
+) -> Epoch {
+    todo!()
+}
+
+pub fn compute_signing_root<T: SimpleSerialize>(
+    ssz_object: T,
+    domain: Domain,
+) -> Result<ssz_rs::Root, ssz_rs::MerkleizationError> {
+    let context = MerkleizationContext::new();
+    let object_root = ssz_object.hash_tree_root(&context)?;
+    let sd = SigningData {
+        object_root,
+        domain,
+    };
+    let context = MerkleizationContext::new();
```

Comment:
ah can re-use this here!

i'm not sure how to communicate this, but i'll go add docs...

for now, it is just read-only

Reply:
a possible future version of this code will actually just pass around a single `MerkleizationContext`

Reply:
fixed